### PR TITLE
(DOCSP-18201): [Swift SDK] Document Type Projection

### PR DESCRIPTION
**Pull request by dacharyc**
### Jira

- https://jira.mongodb.org/browse/DOCSP-18201

### Staged Changes (Requires MongoDB Corp SSO)

- [Type Projection](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-18201/sdk/swift/data-types/type-projection/): New page describing type projection mapping
- [Supported Property Types](https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/DOCSP-18201/sdk/swift/data-types/supported-property-types/#custom-types): Added a new `Type Projection` section with a brief explanation and a link to the new type projection page

### Review Guidelines

[REVIEWING.md](https://github.com/mongodb/docs-realm/blob/master/REVIEWING.md)

### CODE REVIEW

File: source/sdk/swift/data-types/custom-types.txt [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+   These are protocols modeled after Swift's built-in :apple:`RawRepresentable 
+   <documentation/swift/rawrepresentable>`.
+
+.. _ios-persisted-types:
+
+Supported PersistedTypes
+````````````````````````
+
+The ``PersistedType`` can use any of the primitive types that the 
+:ref:`Swift SDK supports <ios-supported-property-types>`. It can also use a
```
Comment (cbush) [inline source/sdk/swift/data-types/custom-types.txt:75]:
```suggestion
:ref:`Swift SDK supports <ios-supported-property-types>`. It can also use an 
```

File: source/sdk/swift/data-types/custom-types.txt [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+   <documentation/swift/rawrepresentable>`.
+
+.. _ios-persisted-types:
+
+Supported PersistedTypes
+````````````````````````
+
+The ``PersistedType`` can use any of the primitive types that the 
+:ref:`Swift SDK supports <ios-supported-property-types>`. It can also use a 
+:ref:`Embedded Object <ios-embedded-objects>` subclass.
```
Comment (cbush) [inline source/sdk/swift/data-types/custom-types.txt:76]: In this case the "subclass" is throwing me off since we're linking to the page about embedded objects in general rather than the specific class (in Swift API docs) `EmbeddedObject`.
```suggestion
:ref:`embedded object <ios-embedded-objects>`.
```

File: source/sdk/swift/data-types/custom-types.txt [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+
+``PersistedType`` cannot be an optional, or a :ref:`collection 
+<ios-client-collections>`. However, the mapped type can be used inside
+an optional or collection.
```
Comment (cbush) [inline source/sdk/swift/data-types/custom-types.txt:80]: This sentence feels unrelated to the current section?

File: source/sdk/swift/data-types/custom-types.txt [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+   the persisted type supports them.
+
+.. literalinclude:: /examples/generated/code/start/CustomTypes.codeblock.query-objects-with-custom-mapped-types.swift
+   :language: swift
+
+Queries on Embedded Objects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can query embedded types on the properties within the object using 
+memberwise equality. However, the you can only query using memberwise
```
Comment (cbush) [inline source/sdk/swift/data-types/custom-types.txt:150]:
```suggestion
memberwise equality. However, you can only query using memberwise 
```

File: source/sdk/swift/data-types/custom-types.txt [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+.. literalinclude:: /examples/generated/code/start/CustomTypes.codeblock.query-objects-with-custom-mapped-types.swift
+   :language: swift
+
+Queries on Embedded Objects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can query embedded types on the properties within the object using 
+memberwise equality. However, the you can only query using memberwise 
+equality if *all* the properties in the embedded object are types which
```
Comment (cbush) [inline source/sdk/swift/data-types/custom-types.txt:151]:
```suggestion
equality if *all* properties in the embedded object are types that 
```

File: source/sdk/swift/data-types/custom-types.txt [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can query embedded types on the properties within the object using 
+memberwise equality. However, the you can only query using memberwise 
+equality if *all* the properties in the embedded object are types which 
+you can compare for equality. Memberwise equality-based querying throws an 
+exception if you cannot compare all properties in the embedded object for 
+equality. 
+
+In this case, passing the mapped type and the persisted type could give
```
Comment (cbush) [inline source/sdk/swift/data-types/custom-types.txt:156]: Passing where? In what case?

File: source/sdk/swift/data-types/supported-property-types.txt [hunk 7]
```diff
@@ -470,6 +470,17 @@ Property Cheat Sheet
...
+------------
+
+.. versionadded:: 10.20.0
+
+You can :ref:`map unsupported types to supported types <ios-custom-types>` 
+to persist in Realm Database. This enables you to work with Swift types 
+that Realm does not support, but store them as types that Realm does support.
+
+You could store a URL as a ``String``, for example, but read it from 
+Realm and use it in your application as though it was a URL.
```
Comment (cbush) [inline source/sdk/swift/data-types/supported-property-types.txt:483]:
```suggestion
Realm and use it in your application as though it were a URL.
```

**Review by cbush — APPROVED** [review]
Looks good. A few questions and typo fixes.

File: source/sdk/swift/data-types/custom-types.txt (hunk 2 above)
Comment (dacharyc) [reply source/sdk/swift/data-types/custom-types.txt:76]: Fair point, changing this to: "can also be an embedded object."

File: source/sdk/swift/data-types/custom-types.txt (hunk 3 above)
Comment (dacharyc) [reply source/sdk/swift/data-types/custom-types.txt:80]: It is related, in that I was trying to point out that while the PersistedType itself can't be an optional, you can use the custom mapped type _as_ an optional in the model. I've added (non-tested) code examples to illustrate. Does that clarify things, or just add length to the page?

File: examples/ios/Examples/CustomTypes.swift [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    convenience init(name: String, url: URL) {
+        self.init()
+        self.id = ObjectId()
+        self.name = name
+        self.url = url
+    }
+
+    convenience init(name: String, url: URL, location: CLLocationCoordinate2D) {
+        self.init()
+        self.id = ObjectId()
```
[the remark this answers is not in the document]
Comment (dacharyc) [reply examples/ios/Examples/CustomTypes.swift:56]: Ugh, I didn't actually need these - I went through several permutations of code trying to understand what I was doing, and missed cleaning these up. Good catch!

File: examples/ios/Examples/CustomTypes.swift [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+// Extend a type as a FailableCustomPersistable if conversion can fail.
+// This returns nil on read if the underlying column contains nil or
+// something that can't be converted to the specified type.
+extension URL: FailableCustomPersistable {
+    public typealias PersistedType = String
+
+    public init?(persistedValue: String) { self.init(string: persistedValue) }
+
+    public var persistableValue: String { self.absoluteString }
```
[the remark this answers is not in the document]
Comment (dacharyc) [reply examples/ios/Examples/CustomTypes.swift:30]: Added comments to explain - does that help?

File: source/sdk/swift/data-types/custom-types.txt [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+Use :swift-sdk:`CustomPersistable <Protocols/CustomPersistable.html>` 
+when there is no chance the conversion can fail.
+
+Use :swift-sdk:`FailableCustomPersistable <Protocols/FailableCustomPersistable.html>` 
+when it is possible for the conversion to fail. 
+
+When a ``FailableCustomPersistable`` property is required, these types are 
+force-unwrapped. If you have a value that can't be converted to the custom 
+type, reading that property throws an unwrapped fail exception.
```
[the remark this answers is not in the document]
Comment (dacharyc) [reply source/sdk/swift/data-types/custom-types.txt]: I moved this down to the section about using custom types in the model - I think it makes more sense there - and rewrote it a bit to lead with the opinionated guidance. 👍

File: source/sdk/swift/data-types/custom-types.txt [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+.. _ios-persisted-types:
+
+Supported PersistedTypes
+````````````````````````
+
+The ``PersistedType`` can use any of the primitive types that the 
+:ref:`Swift SDK supports <ios-supported-property-types>`. It can also use a 
+:ref:`Embedded Object <ios-embedded-objects>` subclass.
+
+``PersistedType`` cannot be an optional, or a :ref:`collection
```
[the remark this answers is not in the document]
Comment (cbush) [reply source/sdk/swift/data-types/custom-types.txt:71]: Resolved but not accepted?

**Review by cbush — APPROVED** [review]

File: source/sdk/swift/data-types/custom-types.txt (hunk 11 above)
[the remark this answers is not in the document]
Comment (dacharyc) [reply source/sdk/swift/data-types/custom-types.txt:71]: Oops! I removed a comma but see this is still here, so not sure if I forgot to save after or removed another poor, innocent comma. Fixing now. Sorry and thank you for catching!

File: examples/ios/Examples/TypeProjection.swift [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+//   }
+// }
+
+import XCTest
+import RealmSwift
+import CoreLocation
+
+// :code-block-start: custom-persistable-protocols
+// Extend a type as a CustomPersistable if the conversion between
+// the mapped type and the persisted type may not fail.
```
Comment (tgoyne) [inline examples/ios/Examples/TypeProjection.swift:13]: "can not fail"? My interpretation of "may not fail" is that if it's not allowed to fail, as opposed to it being impossible for it to fail.

File: source/sdk/swift/data-types/type-projection.txt [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+memberwise equality. However, you can only query using memberwise 
+equality if *all* properties in the embedded object are types which 
+you can compare for equality. Memberwise equality-based querying throws an 
+exception if you cannot compare all properties in the embedded object for 
+equality. 
+
+In this case, querying on the mapped type and the persisted type could give 
+different results. 
+
+.. insert-code-example-here
```
Comment (tgoyne) [inline source/sdk/swift/data-types/type-projection.txt:176]: Missing code sample?

File: source/sdk/swift/data-types/type-projection.txt [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+
+Queries on Embedded Objects
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can query embedded types on the properties within the object using 
+memberwise equality. However, you can only query using memberwise 
+equality if *all* properties in the embedded object are types which 
+you can compare for equality. Memberwise equality-based querying throws an 
+exception if you cannot compare all properties in the embedded object for 
+equality.
```
Comment (tgoyne) [inline source/sdk/swift/data-types/type-projection.txt:171]: This isn't _quite_ true, as object link properties support equality comparisons but aren't supported for memberwise comparisons. It might be more useful to explicitly list which property types are and aren't supported (all primitive types are, objects and collections of any sort aren't).

File: source/sdk/swift/data-types/type-projection.txt [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+you can compare for equality. Memberwise equality-based querying throws an 
+exception if you cannot compare all properties in the embedded object for 
+equality. 
+
+In this case, querying on the mapped type and the persisted type could give 
+different results. 
+
+.. insert-code-example-here
+
+Dynamic APIs
```
Comment (tgoyne) [inline source/sdk/swift/data-types/type-projection.txt:178]: It's probably worth specifically calling out migrations here, as it's the most common use of the dynamic APIs and someone might open this page and ctrl-f migration in the hopes of finding something.
